### PR TITLE
feat(image): add `preview-img-props` prop.

### PR DESCRIPTION
**Pull request by OrbisK**
related to #3515

### CODE REVIEW

**Comment by amadeus711** [comment]
Maybe it's better to call it `img-preview-props`.

File: CHANGELOG.zh-CN.md [hunk 1]
```diff
@@ -55,6 +55,7 @@
 - `n-pagination` 新增 `display-order` 属性，关闭 [#3466](https://github.com/tusen-ai/naive-ui/issues/3466)
 - `n-grid` 新增 `layout-shift-disabled` 属性，关闭 [#3301](https://github.com/tusen-ai/naive-ui/issues/3301)
 - `n-data-table` 新增 `expanded-row-remain-sticky` 属性，支持固定展开内容，关闭 [#3485](https://github.com/tusen-ai/naive-ui/issues/3485).
+- `n-image` add `imgPropsPreview` prop.
```
Comment (amadeus711) [inline CHANGELOG.zh-CN.md:58]:
```suggestion
- `n-image` 新增 `img-props-preview` 属性.
```

**Review by amadeus711 — APPROVED** [review]

**Comment by 07akioni** [comment]
I think `preview-img-props` is a better name?

**Comment by OrbisK** [comment]
> I think `preview-img-props` is a better name?

I've been thinking about that, too. But decided against it, so that no one assumes that these are the props for `preview-image` as component. I am not sure whats better.

**Comment by 07akioni** [comment]
> > I think `preview-img-props` is a better name?
> 
> I've been thinking about that, too. But decided against it, so that no one assumes that these are the props for `preview-image` as component. I am not sure whats better.

I think we can specify it in docs that it's preview image element?

**Comment by OrbisK** [comment]
@07akioni I updated to `preview-img-props` and added a test.

**Comment by OrbisK** [comment]
Rotation of preview overrides `style` 🤔 

Not ready to merge.

**Comment by OrbisK** [comment]
@07akioni not sure what's BP to solve this issue.

**Comment by 07akioni** [comment]
> @07akioni I am not sure what's BP to solve this issue.

Do not let new props override builtin props. If you want it work for style, code like this:

```
<img
  {...imgProps}
  xxx
  style={[
    imgProps?.style,
    { transform: xxx }
  ]}
/>
```

**Comment by OrbisK** [comment]
> > @07akioni I am not sure what's BP to solve this issue.
> 
> Do not let new props override builtin props. If you want it work for style, code like this:
> 
> ```
>    {...imgProps}
>   xxx
>   style={[
>     imgProps?.style,
>     { transform: xxx }
>   ]}
> />
> ```

I guess its not that easy. Problem is that e.g. dragging logic is overriding `templateRef.style` directly.

I had to ref the internal styling. I mean its working.. but I am not sure if this is the BP. I got still a small bug tho. 

### Reprod:
1. Open image docs.
1. Click on image with preview.
1. close preview

### What's happening:
Image is "jumping" when closing

**Comment by 07akioni** [comment]
> 

Yep, I think you should change style to ref, you can derive value using imgProps in `deriveStyle`. I think it may solve the problem.

**Comment by OrbisK** [comment]
Is there any good way to merge `CSSStyleDeclaration` (used to override style in `deriveStyle`) and `StyleValue` (`style` passed as `imgProp`)?

**Comment by 07akioni** [comment]
> Is there any good way to merge `CSSStyleDeclaration` (used to override style in `deriveStyle`) and `StyleValue` (`style` passed as `imgProp`)?

As any is okay.

**Comment by OrbisK** [comment]
> > Is there any good way to merge `CSSStyleDeclaration` (used to override style in `deriveStyle`) and `StyleValue` (`style` passed as `imgProp`)?
> 
> As any is okay.

I am not talking about types. 

`StyleValue` used in `deriveStyle` (`templateRef.value.style`) is not easyly mergable with `CSSStyleDeclaration` passed as `inputProps.style`.

`StyleValue` works like:
```ts
templateRef.value.style.cssText = 'color: red; height: 100px'
// OR
templateRef.value.style.color = 'red'
templateRef.value.style.height = '100px'
```
while `CSSStyleDeclaration` can a more complex type.

I've tried to demonstrate the issue [here](https://sfc.vuejs.org/#eNqFkdFrwjAQxv+VW16qoI2+liobTLYxpiADH5Y9aD1tXZqEJHUbpf/7Lq042WB7Cbn7fvflI1ezG2PiY4UsYanLbGE8OPSVmQpVlEZbDzVY3EEDO6tLiAiNhBIq08p5KN0eJkHvRfcopYaVtnJ7FfW/EY90dIyqpLxQLG5J6PVhMq2FgpaMj2tZYez8p0QSX+pMS20TiAiOmgFEORb73CcwHo3MR/QqVBMMU95lp9RUeCyNXHukCiDNx5C0fhPB/vdj09Vi+fgwv0t5Pj4bUHqaDgEJCE2A+eIZTmhLnfFN5b1WcJ3JInujKXqJhpazW3iapbxTiUz5OSYbsO6vh+XaxAenFW2j/RJxEpxgCbSd0KMdhFqw3HvjEs7dLgs7PLhY2z2nW2wr5YsSY3TlcGP1u0NLxoINLjw4NY9ohxbVFi3avzx/oL98gy1tomHNF/qFx3w=)

**Comment by 07akioni** [comment]
> > > Is there any good way to merge `CSSStyleDeclaration` (used to override style in `deriveStyle`) and `StyleValue` (`style` passed as `imgProp`)?
> > 
> > 
> > As any is okay.
> 
> I am not talking about types.
> 
> `StyleValue` used in `deriveStyle` (`templateRef.value.style`) is not easyly mergable with `CSSStyleDeclaration` passed as `inputProps.style`.
> 
> `StyleValue` works like:
> 
> ```ts
> templateRef.value.style.cssText = 'color: red; height: 100px'
> // OR
> templateRef.value.style.color = 'red'
> templateRef.value.style.height = '100px'
> ```
> 
> while `CSSStyleDeclaration` can a more complex type.
> 
> I've tried to demonstrate the issue [here](https://sfc.vuejs.org/#eNqFkdFrwjAQxv+VW16qoI2+liobTLYxpiADH5Y9aD1tXZqEJHUbpf/7Lq042WB7Cbn7fvflI1ezG2PiY4UsYanLbGE8OPSVmQpVlEZbDzVY3EEDO6tLiAiNhBIq08p5KN0eJkHvRfcopYaVtnJ7FfW/EY90dIyqpLxQLG5J6PVhMq2FgpaMj2tZYez8p0QSX+pMS20TiAiOmgFEORb73CcwHo3MR/QqVBMMU95lp9RUeCyNXHukCiDNx5C0fhPB/vdj09Vi+fgwv0t5Pj4bUHqaDgEJCE2A+eIZTmhLnfFN5b1WcJ3JInujKXqJhpazW3iapbxTiUz5OSYbsO6vh+XaxAenFW2j/RJxEpxgCbSd0KMdhFqw3HvjEs7dLgs7PLhY2z2nW2wr5YsSY3TlcGP1u0NLxoINLjw4NY9ohxbVFi3avzx/oL98gy1tomHNF/qFx3w=)

Use vue's util function:

image

https://sfc.vuejs.org/#eNqFUk2PmzAQ/StTXyASMZsrIlErNWqrqhtpW2kPcQ6ETAKJsa2xoR8R/70DbNNtK3UviOF9+DHPV/HGOdm1KDKR+5JqF8BjaN1KmbpxlgJcgfCYgLHUFLr+gZ/Dd43Qw5FsAxFLI2WUKa3xARp/guXAj6P3qLWFR0v68Cqa/aYE5MfEMa3WzxDCAwPxDJarqzIAm/0ZyyDRBKrRx38GiLfX0mpLGUSsi/oEogrrUxUyWNzduW/RbjaTR0vroqzieHtJoNsNzjBawxhDdoVuUfrBb3vZ8eHdAPacqVcmT6d98CZ4CNg4XQTkCSCvFpCNsqUSLwZRYvW4efj44f5dnlaLmwFvgNVDDiZMoe43X+CJOrJu9H0bgjXwutR1eWEVn8Sih/Vb+LTO0wllZp7eYopETP3Nm8LJs7eGGx7/XT0BXons1zaU4B6HWYkqBOezNPXHcrgXZy8tnVJ+k9SaUDco0TfzPdmvHomNlUieeaT8sUOaE5oDEtL/PP+i/uM7dsFViP4ns9/j8A==

**Comment by OrbisK** [comment]
> > > > Is there any good way to merge `CSSStyleDeclaration` (used to override style in `deriveStyle`) and `StyleValue` (`style` passed as `imgProp`)?
> 
> > > 
> 
> > > 
> 
> > > As any is okay.
> 
> > 
> 
> > I am not talking about types.
> 
> > 
> 
> > `StyleValue` used in `deriveStyle` (`templateRef.value.style`) is not easyly mergable with `CSSStyleDeclaration` passed as `inputProps.style`.
> 
> > 
> 
> > `StyleValue` works like:
> 
> > 
> 
> > ```ts
> 
> > templateRef.value.style.cssText = 'color: red; height: 100px'
> 
> > // OR
> 
> > templateRef.value.style.color = 'red'
> 
> > templateRef.value.style.height = '100px'
> 
> > ```
> 
> > 
> 
> > while `CSSStyleDeclaration` can a more complex type.
> 
> > 
> 
> > I've tried to demonstrate the issue [here](https://sfc.vuejs.org/#eNqFkdFrwjAQxv+VW16qoI2+liobTLYxpiADH5Y9aD1tXZqEJHUbpf/7Lq042WB7Cbn7fvflI1ezG2PiY4UsYanLbGE8OPSVmQpVlEZbDzVY3EEDO6tLiAiNhBIq08p5KN0eJkHvRfcopYaVtnJ7FfW/EY90dIyqpLxQLG5J6PVhMq2FgpaMj2tZYez8p0QSX+pMS20TiAiOmgFEORb73CcwHo3MR/QqVBMMU95lp9RUeCyNXHukCiDNx5C0fhPB/vdj09Vi+fgwv0t5Pj4bUHqaDgEJCE2A+eIZTmhLnfFN5b1WcJ3JInujKXqJhpazW3iapbxTiUz5OSYbsO6vh+XaxAenFW2j/RJxEpxgCbSd0KMdhFqw3HvjEs7dLgs7PLhY2z2nW2wr5YsSY3TlcGP1u0NLxoINLjw4NY9ohxbVFi3avzx/oL98gy1tomHNF/qFx3w=)
> 
> 
> 
> Use vue's util function:
> 
> 
> 
> image
> 
> 
> 
> https://sfc.vuejs.org/#eNqFUk2PmzAQ/StTXyASMZsrIlErNWqrqhtpW2kPcQ6ETAKJsa2xoR8R/70DbNNtK3UviOF9+DHPV/HGOdm1KDKR+5JqF8BjaN1KmbpxlgJcgfCYgLHUFLr+gZ/Dd43Qw5FsAxFLI2WUKa3xARp/guXAj6P3qLWFR0v68Cqa/aYE5MfEMa3WzxDCAwPxDJarqzIAm/0ZyyDRBKrRx38GiLfX0mpLGUSsi/oEogrrUxUyWNzduW/RbjaTR0vroqzieHtJoNsNzjBawxhDdoVuUfrBb3vZ8eHdAPacqVcmT6d98CZ4CNg4XQTkCSCvFpCNsqUSLwZRYvW4efj44f5dnlaLmwFvgNVDDiZMoe43X+CJOrJu9H0bgjXwutR1eWEVn8Sih/Vb+LTO0wllZp7eYopETP3Nm8LJs7eGGx7/XT0BXons1zaU4B6HWYkqBOezNPXHcrgXZy8tnVJ+k9SaUDco0TfzPdmvHomNlUieeaT8sUOaE5oDEtL/PP+i/uM7dsFViP4ns9/j8A==

This is exactly what I was looking for!

File: src/image/src/ImagePreview.tsx [hunk 2]
```diff
@@ -335,6 +338,13 @@ export default defineComponent({
           transformStyle +
           (transition ? '' : 'transition: none;')
       }
+      if (props.imgProps?.style) {
+        Object.entries(
+          normalizeStyle([props.imgProps?.style]) as Record<string, string>
+        ).forEach(([k, v]) => {
+          style[k] = v
+        })
```
Comment (OrbisK) [inline src/image/src/ImagePreview.tsx:346]: image

I dont get the issue. Can someone help?

Comment (07akioni) [reply src/image/src/ImagePreview.tsx:346]: Seems style is typed as `Record<number, xxx>`?

**Comment by 07akioni** [comment]
added in  abd7775

**Pull request closed without merging**